cmake/module/TryAppendCXXFlags.cmake

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+include_guard(GLOBAL)
+include(CheckCXXSourceCompiles)
+
+#[=[
+Usage examples:
+
+  try_append_cxx_flags(warn_cxx_flags "-Wformat -Wformat-security")
+
+
+  try_append_cxx_flags(warn_cxx_flags "-Wsuggest-override"
+    SOURCE "struct A { virtual void f(); }; struct B : A { void f() final; };"
+  )
+
+
+  try_append_cxx_flags(sanitizers_cxx_flags "-fsanitize=${SANITIZERS}" RESULT_VAR cxx_supports_sanitizers)
+  if(NOT cxx_supports_sanitizers)
+    message(FATAL_ERROR "Compiler did not accept requested flags.")
+  endif()
+
+
+  try_append_cxx_flags(warn_cxx_flags "-Wunused-parameter" IF_CHECK_PASSED "-Wno-unused-parameter")
+
+
+  try_append_cxx_flags(error_cxx_flags "-Werror=return-type"
+    IF_CHECK_FAILED "-Wno-error=return-type"
+    SOURCE "#include <cassert>\nint f(){ assert(false); }"
+  )
+
+]=]
+function(try_append_cxx_flags flags_var flags)
+  cmake_parse_arguments(PARSE_ARGV 2
+    TACXXF                            # prefix
+    ""                                # options
+    "SOURCE;RESULT_VAR"               # one_value_keywords
+    "IF_CHECK_PASSED;IF_CHECK_FAILED" # multi_value_keywords
+  )
+
+  string(MAKE_C_IDENTIFIER "${flags}" result)
+  string(TOUPPER "${result}" result)
+  set(result "CXX_SUPPORTS_${result}")
+  unset(${result})
+
+  if(TACXXF_SOURCE)
+    set(source "${TACXXF_SOURCE}")
+    unset(${result} CACHE)
```

Comment:
https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html:

> The check is only performed once, with the result cached in the variable named by `<resultVar>`. Every subsequent CMake run will re-use this cached value rather than performing the check again, even if the `<code>` changes. In order to force the check to be re-evaluated, the variable named by `<resultVar>` must be manually removed from the cache.

We are interested in performing checks whenever the tested source code is changed. This occurs when the `SOURCE` argument is added, omitted after being used before, or modified.

And it has nothing about the `RESULT_VAR` argument.

[Reworked](https://github.com/hebasto/bitcoin/pull/18#issuecomment-1703800481), because the previous branch version had a bug: skipping a check when the `SOURCE` argument is omitted after being used before.